### ISSUE TITLE
Transfer learning SSD

### ISSUE THREAD

**Issue by pinkerltm** [pierluigiferrari/ssd_keras#112]
First thank you a lot for this extraordinary work. This well documented implementation helped me so much (better than any video course out there and even better than my college seminars) to understand concepts and get my hands dirty on deep learning and especially training detectors.

I am using this detector for my research on transfer learning for traffic sign detection. The Point is, that a well trained traffic sign detector only works in the Country the Training data was originated. If you plan to use the detector in another Country (with slightly different looking signs) you possibly want to fine tune or better transfer learn on a couple of training Images (possibly you don't have very much of them ~100). Also my intial Training set, ~5000 Traffic sign Images from belgium aren't that much, so that it might be wise to Transfer learn from MS COCO for example.

I also map traffic signs to 80 classes according to the MS COCO Class Count. I loaded the pretrained COCO weights from your site, freezed all Convolutional layers of the VGG16 Basenetwork and started to Train, expecting that this should be much faster, than Training from scratch. 

But the training process didn't speed up in any kind. And also loss was only slowing converging, just as if I am training from scratch. Finally it didn't go down as far as the model I trained from scratch.

Is it even possible to Transfer learn on SSD this way? What to do with the mbox layers? Is there also a Need to manipulate the weights or freeze them? Could there be a problem with Keras or Tensorflow and this particular implementation which could make it not working?

Sorry if I missed any part of the documentation, I carefully read every part I was thinking should be involved in this process.

**Comment by pierluigiferrari** [comment]
Thanks, I'm happy that this is useful for others. As for your questions:

> I loaded the pretrained COCO weights from your site, freezed all Convolutional layers of the VGG16 Basenetwork

I wouldn't freeze any weights ever. I don't understand why people always do that when they perform transfer learning. Sure, freezing some layers might speed up the training a tiny bit, but if it were more optimal with regard to my fine-tuning dataset if the weights of any layer were different, I would certainly want the training to adjust these weights rather than them being frozen, otherwise I'm just unnecessarily capping the maximal accuracy the model can reach on my dataset.

> But the training process didn't speed up in any kind. And also loss was only slowing converging, just as if I am training from scratch.

How long it takes for the training to converge depends on how "far" the initial weights are from the target weights. As such, it cannot be generally said how long it should take. You can only say that the closer the weights you start out with are to the theoretically ideal weights for your dataset, the faster the training should converge. This also means that the more different your dataset is from the dataset that the pre-trained model was trained on, the longer the training will take to converge.

> Is it even possible to Transfer learn on SSD this way?

I'm not sure what you mean by "this way", but of course it's possible to use a trained SSD for transfer learning.

> What to do with the mbox layers?

I don't understand the question.

>  Is there also a Need to manipulate the weights or freeze them?

Do you mean the `mbox` layers? They don't have weights.

> Could there be a problem with Keras or Tensorflow and this particular implementation which could make it not working?

If there was a problem with Keras, TensorFlow, or this particular implementation, it would very likely not be possible for this SSD implementation to reproduce the original SSD results, so it's unlikely that this is the case. It's more likely that something about the particular transfer learning setup isn't right.

**Comment by pinkerltm** [comment]
ad freezing layers: Well the purpose would be a drastic increase of learning speed, as there are not 33 Mio. Parameters to Train as in my current model, but only (let's say) 5 Mio. which should speed up things not a tiny bit, but by a significant factor. The other aspect is the fear of the "Catastrophic forgetting" daemon I guess...but I have to admit that I also think that if the model really "forgets" features which aren't very common in my data anyway, I would not bother about it.

ad Training time: Ok, but that would mean that the random initial weights for VGG16 are closer to my use case than weights trained on COCO or Deepstreet. Thats interesting, I'll try to verify that because I think that would be worth to get published in my Thesis.

ad "this way": Thanks that answers my question...I just wanted to be secure that Transfer learning on SSD just includes loading pretrained weights, freeze (or not) feature layers and to adopt the finaly detection layer accordingly to the new set of classes

ad "mbox" layers: Sorry for being unclear, I meant the bounding box regression layers after the base classification Network (so conv5_* and pool5 for VGG16). As there are also convolutional layers, I thought there also might be weights which should be given any initial weight which makes sense. Am I able to load the initial weights, and then overload the VGG16 weights with weights from a pure VGG16 classification Network like Deepstreet?

**Comment by pinkerltm** [comment]
> why would one want to artificially limit the best achievable state by freezing weights just to get there faster?
Because you are possibly limited by:

- Hardware (GPU, Memory, etc...) ... usually I am processing on my Geforce GTX1070, while I can only dream of Clusters of teslas computing weights of hundrets of ResNet Layers

- Data ... I only have a couple of Training samples to adopt to...I don't want to train my classificator over and over again on the same 50 Images just to fine tune it on new conditions, when my training data to get to these weights and features had several thousand training examples

- Time to get productive ... my Research is on the Topic of Training a traffic sign detector on a rather large dataset of one Country, and then reuse it in another Country (with different looking traffic signs)... the detector should be adopted in the fastest (most efficient) manner.

But you are right, if I don't need to care about These limitations, I would pour all data Images into one big pot and train on just this until the best achievable state comes out. 

> In order to draw such a conclusion you would have to be really certain that it's correct.
You are right, I sorted it out...there were two Major mistakes in my Training Notebook. 

1. Training on the traffic signs doesn't converge very well on the scales for the COCO dataset. On the other Hand it does converge very well on the Pascal scales. I used the COCO scales on my second experiment because I also used pretrained COCO weights instead of random weights, this seems to not fit on my data very well. I will stick on PASCAL scales in future experiments and try to tweak them a Little bit to better fit my data

1. After dropping the "Background" class from my data, the model converges much better.

Thanks for all the clarifications, I think this can be closed now.
